A17-client-side-health-checking.md

```diff
@@ -0,0 +1,315 @@
+Client-Side Health Checking
+----
+* Author(s): Mark D. Roth (roth@google.com)
+* Approver: a11r
+* Status: Draft
+* Implemented in: 
+* Last updated: 2018-08-27
+* Discussion at: https://groups.google.com/d/topic/grpc-io/rwcNF4IQLlQ/discussion
+
+## Abstract
+
+This document proposes a design for supporting application-level
+health-checking on the client side.
+
+## Background
+
+gRPC has an existing [health-checking
+mechanism](https://github.com/grpc/grpc/blob/master/doc/health-checking.md),
+which allows server applications to signal that they are not healthy
+without actually tearing down connections to clients.  This is
+used when (e.g.) a server is itself up but another service that it depends
+on is not available.
+
+Currently, this mechanism is used in some [look-aside
+load-balancing](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md)
+implementations to provide centralized health-checking of backend
+servers from the balancer infrastructure.  However, there is no
+support for using this health-checking mechanism from the client side,
+which is needed when not using look-aside load balancing (or when
+falling back from look-aside load balancing to directly contacting
+backends when the balancers are unreachable).
+
+### Related Proposals
+
+N/A
+
+## Proposal
+
+The gRPC client will be configured to send health-checking RPCs to
+each backend that it is connected to.  Whenever a backend responds as
+unhealthy, the client's LB policy will stop sending requests to that
+backend until it reports healthy again.
+
+Note that because the health-checking service requires a service name, the
+client will need to be configured with a service name to use.  However,
+by default, it can use the empty string, which would mean that the
+health of all services on a given host/port would be controlled with a
+single switch.
+
+### Watch-Based Health Checking Protocol
+
+The current health-checking protocol is a unary API, where the client is
+expected to periodically poll the server.  That was sufficient for use
+via UHC, where the health checks come from a small number of centralized
+clients and where there is existing infrastructure to periodically
+poll each client.  However, if we are going to have large numbers of
+clients issuing health-check requests, then we will need to convert the
+health-checking protocol to a streaming Watch-based API for scalability
+and bandwidth-usage reasons.
+
+Note that one down-side of this approach is that it could conceivably
+be possible that the server-side health-checking code somehow fails
+to send an update when becoming unhealthy.  If the problem is due to
+a server that stops polling for I/O, then the problem would be caught
+by keepalives, at which point the client would disconnect.  But if the
+problem is caused by a bug in the health-checking service, then it's
+possible that a server could still be responding but has failed to notify
+the client that it is unhealthy.
```

Comment:
You could also advertise the server's status via trailer metadata. So active RPCs that are issued while the service is unhealthy could attach the trailer and client could learn of the server's change in disposition that way, as a fallback (in case server failed to deliver update on main health-checking stream). This would not rely on polling, and HTTP/2 header compression should be able to encode the value efficiently if it is included in trailer for every RPC, so overhead would be very low. So if a channel were idle and became unhealthy (and server failed to deliver update), the client would not discover this until its first RPC on that channel.

Reply:
That could certainly work, but it seems like a lot of overhead in terms of both code complexity (there would be a lot of plumbing involved) and performance (the cost of encoding and decoding the metadata -- it turns out that from a performance perspective, HTTP/2 header compression is actually one our biggest expenses).  I'm not sure it's really worthwhile for something that is a very unlikely failure mode to begin with.

Reply:
@markdroth, I'm not sure what performance cost your concerned about since the header would only need to exist when there's a problem. The bigger questions in my mind are 1) how do we know the "scope" of the failure trailer, that is, which services/methods are impacted and 2) how do we know when the service improves?

It does also require issuing an RPC to a unhealthy backend before you learn it is unhealthy, but depending on who you talk to that's more minor.

Reply:
You may be right that the performance isn't a big deal.  But the code complexity still seems significant.  I'm just not sure it's worth this much trouble to address something that's extremely unlikely to begin with.